### PR TITLE
Basic sybill resistance mechanisms

### PR DESCRIPTION
**Pull request by whyrusleeping**
This needs a bit more work still, but it should help.

Things we should do before merging:
- [x] Fix passing the protected peers from the config to the connmgr
- [x] Add config entries for other connmgr settings (hi, lo, grace)

We should deploy the bootstrappers with a much higher connection limit, and a pretty reasonable grace period (1 minute?).  We should also add the peer IDs of our miners (miners daemons) and our bootstrappers to the protected peers in all our nodes configs.

### CODE REVIEW

File: chain/blocksync/blocksync_client.go [hunk 1]
```diff
@@ -264,6 +269,10 @@ func (bs *BlockSync) sendRequestToPeer(ctx context.Context, p peer.ID, req *Bloc
 }
 
 func (bs *BlockSync) processBlocksResponse(req *BlockSyncRequest, res *BlockSyncResponse) ([]*types.TipSet, error) {
+	if len(res.Chain) == 0 {
```
Comment (sternhenri) [inline chain/blocksync/blocksync_client.go]: this looks good. If it doesn't end up doing it, next step would be to set a target based on best peer performance (could be an attack vector but i think it's ok to assume at least one honest peer and update to not number of blocks sent but number of blocks verified in best chain if needed (quite a bit more complex))

File: node/config/def.go [hunk 2]
```diff
@@ -67,6 +72,10 @@ func defCommon() Common {
 				"/ip4/0.0.0.0/tcp/0",
 				"/ip6/::/tcp/0",
 			},
+
```
Comment (magik6k) [inline node/config/def.go]: Do we maybe want to protect bootstrappers by default?

File: chain/blocksync/blocksync_client.go [hunk 3]
```diff
@@ -101,8 +101,13 @@ func (bs *BlockSync) GetBlocks(ctx context.Context, tsk types.TipSetKey, count i
 		}
 
 		if res.Status == 0 {
+			resp, err := bs.processBlocksResponse(req, res)
+			if err != nil {
+				return nil, xerrors.Errorf("success response from peer failed to process: %w", err)
+			}
 			bs.syncPeers.logGlobalSuccess(time.Since(start))
-			return bs.processBlocksResponse(req, res)
+			bs.host.ConnManager().TagPeer(p, "bsync", 2)
```
Comment (Stebalien) [inline chain/blocksync/blocksync_client.go]: FYI, DHT peers will have a higher priority (`5`, or `5 + log2(xor distance)` when you next update).

File: node/peers/miners.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	for _, m := range mactors {
+		mpow, _, err := stmgr.GetPower(context.TODO(), pt.st, ts, m)
+		if err != nil {
+			log.Warn("failed to get miners power: ", err)
+			continue
+		}
+		if !mpow.IsZero() {
+			pid, err := stmgr.GetMinerPeerID(context.TODO(), pt.st, ts, m)
```
Comment (whyrusleeping) [inline node/peers/miners.go]: Oh, this whole thing doesnt work. This is the peer ID of the storage-miner process, but the storage miner process doesnt participate in block propagation. So forming and keeping connections to them doesnt really help us at all...

File: node/config/def.go (hunk 2 above)
Comment (whyrusleeping) [reply node/config/def.go]: That would be good, bootstrappers can still disconnect from you, but you probably don't ever want to willingly disconnect from them?

File: node/peers/miners.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+			if err != nil {
+				log.Warn("failed to get peer ID for miner: ", err)
+				continue
+			}
+			pt.h.ConnManager().TagPeer(pid, "miner", 10)
```
Comment (Stebalien) [inline node/peers/miners.go]: Even if we're not connected? We'll temporarily tag the peer then forget this. I guess that's intentional?

I'd recommend at least commenting on this behavior in the code.

File: node/peers/miners.go (hunk 4 above)
Comment (magik6k) [reply node/peers/miners.go]: Uh..

File: chain/blocksync/blocksync_client.go (hunk 3 above)
Comment (Stebalien) [reply chain/blocksync/blocksync_client.go]: Also note: Bitswap adds a pretty significant weight by default (even for peers _downloading_ from us). We may need to tune that.

**Pull request merged**